File: hw/tipan/u3/hw28Strategy/StrategyJavaScript/model/QuickSort.js
const SortingStrategy = require("./SortingStrategy");

class QuickSort extends SortingStrategy {

    sort(data) {

        let array = [...data];

        this.quickSort(array, 0, array.length - 1);

        return array;

    }

    quickSort(array, low, high) {

        if (low < high) {

            let pivot = this.partition(array, low, high);

            this.quickSort(array, low, pivot - 1);
            this.quickSort(array, pivot + 1, high);

        }

    }

    partition(array, low, high) {

        let pivot = array[high];
        let i = low - 1;

        for (let j = low; j < high; j++) {

            if (array[j] <= pivot) {

                i++;

                let temp = array[i];
                array[i] = array[j];
                array[j] = temp;

            }

        }

        let temp = array[i + 1];
        array[i + 1] = array[high];
        array[high] = temp;

        return i + 1;

    }

}

module.exports = QuickSort;

File: exams/codena/u3/Strategy/src/main/java/ec/edu/espe/strategy/controller/QuickSort.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.espe.strategy.controller;

import java.util.ArrayList;

/**
 *
 * @author Daniel Codena, @CodeBreakers, ESPE
 */
public class QuickSort extends SortingStrategy {

    @Override
    public ArrayList<Object> sort(ArrayList<Object> numbers) {
        System.out.println("Quick Sort");
        quickSort(numbers, 0, numbers.size() - 1);
        return numbers;
    }

    private void quickSort(ArrayList<Object> numbers, int low, int high) {
        if (low < high) {
            int pi = partition(numbers, low, high);

            quickSort(numbers, low, pi - 1);
            quickSort(numbers, pi + 1, high);
        }
    }

    private int partition(ArrayList<Object> numbers, int low, int high) {

        int pivot = (Integer) numbers.get(high);
        int i = low - 1;

        for (int j = low; j < high; j++) {

            if ((Integer) numbers.get(j) <= pivot) {
                i++;

                Object temp = numbers.get(i);
                numbers.set(i, numbers.get(j));
                numbers.set(j, temp);
            }
        }

        Object temp = numbers.get(i + 1);
        numbers.set(i + 1, numbers.get(high));
        numbers.set(high, temp);

        return i + 1;
    }

}